// ngx-image-compress/esm2022/lib/ngx-image-capture.module.mjs
import { NgModule } from '@angular/core';
import { NgxImageCaptureComponent } from './ngx-image-capture.component';
import { CommonModule } from '@angular/common';
import * as i0 from "@angular/core";
export class NgxImageCaptureModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.1.1", ngImport: i0, type: NgxImageCaptureModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "18.1.1", ngImport: i0, type: NgxImageCaptureModule, declarations: [NgxImageCaptureComponent], imports: [CommonModule], exports: [NgxImageCaptureComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "18.1.1", ngImport: i0, type: NgxImageCaptureModule, imports: [CommonModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.1.1", ngImport: i0, type: NgxImageCaptureModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgxImageCaptureComponent],
                    imports: [CommonModule],
                    exports: [NgxImageCaptureComponent],
                }]
        }] });
